syft/frameworks/torch/linalg/lr.py

```diff
@@ -284,3 +285,247 @@ def _compute_pvalues(self):
+    It uses Secured Multi-Party Computation at combine phase.
+    While the training is performed in SMPC, the final regression coefficients
+    are public at the end.
+
+    Reference: Section 2 of https://arxiv.org/abs/1901.09531
+
+    Args:
+        crypto_provider: a BaseWorker providing crypto elements for ASTs such as
+            Beaver triples
+        hbc_worker: The "Honest but Curious" BaseWorker
+        precision_fractional: precision chosen for FixedPrecisionTensors
+
+    Attributes:
+        coef: torch.Tensor of shape (n_features, ). Estimated coefficients for
+            DASH algorithm.
+        pvalue: numpy.array of shape (n_features, ). Two-sided p-value for a
+            hypothesis test whose null hypothesis is that the each coeff is zero.
+
+    """
+
+    def __init__(
+        self, crypto_provider: BaseWorker, hbc_worker: BaseWorker, precision_fractional: int = 6
+    ):
+
+        self.crypto_provider = crypto_provider
+        self.hbc_worker = hbc_worker
+        self.precision_fractional = precision_fractional
+
+    def fit(
+        self, X_ptrs: List[torch.Tensor], C_ptrs: List[torch.Tensor], y_ptrs: List[torch.Tensor]
+    ):
+
+        # Checking if the pointers are as expected
+        self._check_ptrs(X_ptrs, C_ptrs, y_ptrs)
+
+        # Check if each y is a 2-dim or 1-dim tensor, unsqueeze it if it's 1-dim
+        for i, y in enumerate(y_ptrs):
+            if len(y.shape) < 2:
+                y_ptrs[i] = y.unsqueeze(1)
+
+        self.workers = self._get_workers(X_ptrs)
+
+        # Computing aggregated pairwise dot products remotelly
+        XX_ptrs, Xy_ptrs, yy_ptrs, CX_ptrs, Cy_ptrs = self._remote_dot_products(
+            X_ptrs, C_ptrs, y_ptrs
+        )
+
+        # Compute remote QR decompositions
+        R_ptrs = self._remote_qr(C_ptrs)
+
+        # Secred share tensors between hbc_worker, crypto_provider and a random worker
+        # and compute aggregates. It corresponds to the Combine stage of DASH's algorithm
+        idx = random.randint(0, len(self.workers) - 1)
+        XX_shared = sum(self._share_ptrs(XX_ptrs, idx)).sum(dim=0)
+        Xy_shared = sum(self._share_ptrs(Xy_ptrs, idx))
+        yy_shared = sum(self._share_ptrs(yy_ptrs, idx))
+        CX_shared = sum(self._share_ptrs(CX_ptrs, idx))
+        Cy_shared = sum(self._share_ptrs(Cy_ptrs, idx))
+        R_shared = torch.cat(self._share_ptrs(R_ptrs, idx), dim=0)
+
+        # QR decomposition of R_shared
+        _, R = qr(R_shared, norm_factor=self.total_size ** (1 / 2))
+
+        # Compute inverse of upper matrix
+        R_inv = self._inv_upper(R)
+
+        Qy = R_inv.t() @ Cy_shared
+        QX = R_inv.t() @ CX_shared
+
+        denominator = XX_shared - (QX ** 2).sum(dim=0)
+        # Need the line below to perform inverse of a number in MPC
+        inv_denominator = ((0 * denominator + 1) / denominator).squeeze()
+
+        coef_shared = (Xy_shared - QX.t() @ Qy).squeeze() * inv_denominator
```

Comment:
Can't you spare the line `inv_denominator = ...` by doing
```python
coef_shared = (
 (Xy_shared - QX.t() @ Qy) / denominator
).squeeze()
```

Reply:
if you look at the computations of `coef_shared` and `sigma2_shared` I need to do 2 divisions by the same denominator.

By computing `inv_denominator` and using multiplication instead of division to compute `coef_shared` and `sigma2_shared` I optimize the time execution by having only one division overall.

Reply:
That's right, my bad!